migration/src/spreadsheets.rs

```diff
@@ -137,49 +137,37 @@ impl SheetResult {
     /// has a list of associated Chapters.
     pub fn into_collection_index(self,
         self_title: &String, 
-        self_wordpress_menu_id: &i64) -> Result<Collection> {
+        self_wordpress_menu_id: &i64,
+        self_slug: &String) -> Result<Collection> {
         let mut chapters = Vec::new();
         let mut self_intro_chapters = Vec::new();
-        for row in self.values.into_iter().skip(2){
-            if row[0].is_empty(){
-                self_intro_chapters = chapters.clone();
+        let mut row = self.values.into_iter();
+        let first_value = row.next().ok_or_else(|| anyhow::format_err!("Missing first value"))?;
+        let second_value = row.next().ok_or_else(|| anyhow::format_err!("Missing second value"))?;
+        for cur_row in row{
+            if cur_row[0].is_empty(){
+                self_intro_chapters = chapters;
                 chapters = Vec::<Chapter>::new();
             }
             else {
-                let index_in_parent_string = row[0].clone();
-                let index_i64 = index_in_parent_string.parse::<i64>().unwrap();
+                // Return error if no index in parent is specified.
+                let index_i64 = (&cur_row[0]).parse::<i64>()?;
 
-                let document_if_applicable = row[3].clone();
-                let mut doc_string= None;
-
-                let wp_id_if_applicable = row[5].clone();
-                let mut wp_id = None;
-
-                if wp_id_if_applicable.is_empty(){
-                    wp_id = None;
-                }
-                else {
-                    wp_id = Some(wp_id_if_applicable.parse::<i64>().unwrap());
-                }
-
-                if document_if_applicable.is_empty() {
-                    doc_string = None;
-                }
-                else {
-                    doc_string = Some(document_if_applicable);
-                }
+                // Both of these fields are optional
+                let wp_id = (&cur_row[4]).parse::<i64>().ok();
+                let doc_string = Some(cur_row[5].to_string());
 
                 let new_chapter = Chapter {
                     index_in_parent: index_i64,
-                    url_slug: row[1].clone(),
-                    chapter_name: row[2].clone(),
-                    document_title: doc_string,
+                    url_slug: (&cur_row[1]).to_string(),
+                    chapter_name: (&cur_row[2]).to_string(),
```

Comment:
Hmm doing `(&x).to_string()` is the same as doing `x.clone()` but with more steps, they both allocate. You can't move values out of a `Vec` using the indexing syntax `list[i]`, that's always read-only.  You could do exactly what you did earlier in this function:

```rs
let row_values = cur_row.into_iter();
// you could also use .ok_or_else(...)
let index: i64 = row_values.next().unwrap().parse()?;
let url_slug = row_values.next().unwrap();
// and so forth
```

This and your current code will panic if a row doesn't have enough values. You should check at the beginning `cur_row.len() > 4` or however many values it needs, then using `.unwrap()` or `cur_row[i]` here is pretty safe.

Reply:
okay-- for some reason I couldn't get it to work using a check for cur_row.len() (it compiled and everything, the right values just weren't showing up in the tables). Instead, I used a peekable iterator before using .unwrap(). Would that be an okay solution?

Reply:
Hmm that might work fine, does it let you peek multiple cells or something?

Reply:
The last three fields of the spreadsheet are optional, so I just used peek three times before assigning data to any of those fields. (I also committed files with this change if you wanted to take a look)